kernel.c

```diff
@@ -0,0 +1,728 @@
+    (((int)cp -
+      ((argcount + 3 + POST_ARGV_NULL_SPACE) *sizeof (void *)))
+     & ~7);
+
+  /*
+   * Compute the new stack pointer, leaving INITIAL_STACK_FRAME_SIZE bytes
+   * reserved above the stack pointer, before the arguments.
+   */
+  cp2 = (caddr_t)cpp - INITIAL_STACK_FRAME_SIZE;
+
+
+
+  TracePrintf(1, "prog_size %d, text %d data %d bss %d pages\n",
+              li.t_npg + data_npg, li.t_npg, li.id_npg, li.ud_npg);
+
+
+  /*
+   * Compute how many pages we need for the stack */
+  stack_npg = (VMEM_1_LIMIT - DOWN_TO_PAGE(cp2)) >> PAGESHIFT;
+
+  TracePrintf(1, "LoadProgram: heap_size %d, stack_size %d\n",
+              li.t_npg + data_npg, stack_npg);
+
+
+  /* leave at least one page between heap and stack */
+  if (stack_npg + data_pg1 + data_npg >= MAX_PT_LEN) {
+    close(fd);
+    return ERROR;
+  }
+
+  /*
+   * This completes all the checks before we proceed to actually load
+   * the new program.  From this point on, we are committed to either
+   * loading succesfully or killing the process.
+   */
+
+  /*
+   * Set the new stack pointer value in the process's UserContext
+   */
+
+  /*
+   * ==>> (rewrite the line below to match your actual data structure)
+   * ==>> proc->uc.sp = cp2;
+   */
+  proc->uctxt.sp = cp2;
+
+  /*
+   * Now save the arguments in a separate buffer in region 0, since
+   * we are about to blow away all of region 1.
+   */
+  cp2 = argbuf = (char *)malloc(size);
+
+  /*
+   * ==>> You should perhaps check that malloc returned valid space
+   */
+
+  if (argbuf == NULL) {
+    TracePrintf(0, "Failed to allocate memory for argbuf\n");
+    Halt();
+  }
+
+  for (i = 0; args[i] != NULL; i++) {
+    TracePrintf(3, "saving arg %d = '%s'\n", i, args[i]);
+    strcpy(cp2, args[i]);
+    cp2 += strlen(cp2) + 1;
+  }
+
+  /*
+   * Set up the page tables for the process so that we can read the
+   * program into memory.  Get the right number of physical pages
+   * allocated, and set them all to writable.
+   */
+
+  /* ==>> Throw away the old region 1 virtual address space by
+   * ==>> curent process by walking through the R1 page table and,
+   * ==>> for every valid page, free the pfn and mark the page invalid.
+   */
+
+  for (int i = 0; i <MAX_PT_LEN; i++){
+  
+    if (proc->region1_pt[i].valid){
+    
+      free_frame_number(proc->region1_pt[i].pfn);
+
+      proc->region1_pt[i].valid = 0;
+    
+    }
+  
+  }
+
+  /*
+   * ==>> Then, build up the new region1.
+   * ==>> (See the LoadProgram diagram in the manual.)
+   */
+
+  /*
+   * ==>> First, text. Allocate "li.t_npg" physical pages and map them starting at
+   * ==>> the "text_pg1" page in region 1 address space.
+   * ==>> These pages should be marked valid, with a protection of
+   * ==>> (PROT_READ | PROT_WRITE).
+   */
+
+  int text_top = text_pg1+ li.t_npg;
+
+
+  for (int i = text_pg1; i < text_top; i++){
+  
+    int index = get_free_frame();
+
+    if (index < 0){
+    
+      for (int i = text_pg1; i < text_top; i++){
```

Comment:
nit: please include comments to explain what you're doing here (i.e. aborting & freeing)

Reply:
another small nit: consider using the `ERROR` macro from `ykernel.h` rather than hardcoded `-1`

Reply:
also, since you know the number of free frames the program will need (by adding the text, data, and stack pages), you could figure out if you'll have enough memory before comitting to the actual program load (by counting the number of free physical frames and comparing), rather than finding out as you're trying to load the program that "opps, we ran out of memory". 

This could allow a more for more fault tolerance -- for example, a userland program could call `exec` and then actually have backup code if that exec fails.